tests/integration/test_issue_2412.py

```diff
@@ -0,0 +1,257 @@
+    create_entry_module(project_dir)
+    input_requirements = os.path.join(project_dir, "requirements.in")
+    with safe_open(input_requirements, "w") as fp:
+        print("numpy", file=fp)
+    create_setup_py(project_dir)
+    return project_dir
+
+
+def assert_bdist_pex(
+    project_dir,  # type: str
+    expected_numpy_version,  # type: str
+):
+    # type: (...) -> None
+
+    assert "Hello  {numpy_version}\n".format(
+        numpy_version=expected_numpy_version
+    ) == subprocess.check_output(args=[os.path.join(project_dir, "dist", "entry")]).decode("utf-8")
+
+
+skip_if_incompatible_with_repro_project = pytest.mark.skipif(
+    not InterpreterConstraint.matches("CPython>=3.9,<3.13"),
+    reason=(
+        "The repro project under test requires Python ~=3.9 and we further restrict to CPython "
+        "and place a version ceiling at 3.13 to ensure we resolve numpy wheels and do not need to "
+        "build the sdist."
+    ),
+)
+
+
+@skip_if_incompatible_with_repro_project
+def test_bdist_pex_locked_issue_2412_repro_exact(
+    issue_2412_repro_project_dir,  # type: str
+    pex_project_dir,  # type: str
+):
+    # type: (...) -> None
+
+    venv = Virtualenv.create(
+        venv_dir=os.path.join(issue_2412_repro_project_dir, ".env"),
+        install_pip=InstallationChoice.UPGRADED,
+        install_setuptools=InstallationChoice.UPGRADED,
+    )
+    subprocess.check_call(
+        args=[venv.bin_path("pip"), "install", "-U", "pip-tools", pex_project_dir]
+    )
+
+    subprocess.check_call(
+        args=[
+            venv.bin_path("pip-compile"),
+            "--output-file",
+            "requirements.txt",
+            "--generate-hashes",
+            "requirements.in",
+        ],
+        cwd=issue_2412_repro_project_dir,
+    )
+    locked_requirements = list(
+        parse_requirement_file(os.path.join(issue_2412_repro_project_dir, "requirements.txt"))
+    )
+    assert 1 == len(locked_requirements)
+
+    locked_requirement = locked_requirements[0]
+    assert isinstance(locked_requirement, PyPIRequirement)
+    assert (
+        "--hash=sha256:" in locked_requirement.line.raw_text
+    ), "Expected the compiled requirements file to include hashes."
+
+    locked_numpy_specifiers = list(locked_requirement.requirement.specifier)
+    assert 1 == len(locked_numpy_specifiers)
+    locked_numpy_specifier = locked_numpy_specifiers[0]
+    assert "==" == locked_numpy_specifier.operator
+    locked_numpy_version = locked_numpy_specifier.version
+
+    subprocess.check_call(
+        args=[
+            venv.interpreter.binary,
+            "setup.py",
+            "bdist_pex",
+            "--pex-args",
+            "--disable-cache -vvvv -r requirements.txt --pip-version 24.0",
+            "--bdist-all",
+        ],
+        cwd=issue_2412_repro_project_dir,
+    )
+    assert_bdist_pex(issue_2412_repro_project_dir, expected_numpy_version=locked_numpy_version)
```

Comment:
Alright @anisse, now your exact use case works. I added alternatives below that use Pex locking instead of `pip-compile`. I will note that `pex3 lock create --style universal --interpreter-constraint <matching your project> ...` does not have the limitations described here for `pip-compile`: https://pip-tools.readthedocs.io/en/stable/#cross-environment-usage-of-requirements-in-requirements-txt-and-pip-compile The Pex `--style universal` lock is truly universal, including all possibly needed depepdency subgraphs and resolving the right one when the lock is consumed later on a target machine to create a PEX for that machine. Or even when identifying the (foreign) target machine with `--platform` or `--complete-platform` to build, for example, a Linux PEX on a Mac or vice-versa or even a Linux / Mac multi-platform PEX.